### PR TITLE
[5.2] Model accessors used in custom relationship columns

### PR DESCRIPTION
**Pull request by kirkbushell**
There is a bug that exists whereby if a model's accessor returns a value other than primitive (such as an array, due to json decoding.etc.), and that same column is used for a join, then the relationship completely fails as an array cannot be used as part of the relationship value.

This PR resolves that by forcing the value to a string, which is great for things like value objects, but should your column return an array or some other object, this will return a value that will still allow the relationship to be resolved.

This is especially useful if values can consist of more than one return type - like a string, and/or array (think custom field values).

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Any chance of a test being added please?

**Comment by kirkbushell** [comment]
@GrahamCampbell I don't think a test is needed here. I'd literally be testing PHP functionality. There's no additional business logic or complexity as a result of this PR.

**Comment by GrahamCampbell** [comment]
You need to test that proves you actually need a cast here, otherwise there is no argument for making the change.

**Comment by kirkbushell** [comment]
Okay that's a good point - I'll add one in to demonstrate the issue.

**Comment by kirkbushell** [comment]
@GrahamCampbell right, so that manages it - however I think some discussion needs to be had around handling other values. Technically, in the case of array/object.etc. it could create problems in returning a record it shouldn't, so I think maybe even serializing the variable instead of getting the type may be better?

The desired outcome is to have a string return value.

File: src/Illuminate/Database/Eloquent/Relations/Relation.php [hunk 1]
```diff
@@ -199,7 +199,17 @@ public static function noConstraints(Closure $callback)
...
-            return $key ? $value->getAttribute($key) : $value->getKey();
+            if ($key) {
+                $value = $value->getAttribute($key);
+
+                if (is_scalar($value)) {
+                    return $value;
+                }
+
+                return gettype($value);
+            }
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/Relation.php]: seems a bit odd

**Comment by taylorotwell** [comment]
I dunno, this seems odd. I don't think we want foreign key accessors returning things like "StdClass" or something like that and trying to query on it.

**Comment by kirkbushell** [comment]
I was hoping for more of a discussion on it. Currently were unable to use eager loading because of his issue in a particular use-case. I agree returning the type is odd (and problematic), but is there another solution we can work towards?

**Comment by kirkbushell** [comment]
@taylorotwell - another idea is to serialize the data or something. It's odd, to be sure - but I'm not sure how else to handle it at the moment.

**Comment by kirkbushell** [comment]
I feel we should find a solution here, as even casting supported by the library will cause a similar problem if you cast to an array - and it's very hard to debug.

**Pull request closed without merging**